### PR TITLE
feat: optimize cache

### PR DESCRIPTION
**Pull request by DjDeveloperr**
- adds optimized .size methods to managers and child managers
- adds .memberCacheSize to GuildsManager and .messageCacheSize to ChannelsManager
- fixes user not being cached from member object and duplication of user object in member objects
- added per channel message cache limit and corresponding option
- added option to disable compression

there are some breaking changes for cache adapter too, though

tested bot on 300 guilds
before: 100mb+ usage
now: 20mb~ usage

### CODE REVIEW

File: src/managers/memberRoles.ts [hunk 1]
```diff
@@ -39,6 +39,12 @@ export class MemberRolesManager extends BaseChildManager<RolePayload, Role> {
     else return undefined
   }
 
+  async size(): Promise<number> {
+    const payload = await this.member.guild.members._get(this.member.id)
+    if (payload === undefined) return 0
+    else return payload.roles.length
```
Comment (MierenManz) [inline src/managers/memberRoles.ts:45]: `else` is not needed in this case

File: src/managers/messages.ts [hunk 2]
```diff
@@ -24,24 +25,30 @@ export class MessagesManager extends BaseManager<MessagePayload, Message> {
...
 
     if (author === undefined) author = new User(this.client, raw.author)
 
-    const res = new this.DataType(this.client, raw, channel, author) as any
+    const res = new this.DataType(this.client, raw, channel, author)
     await res.mentions.fromPayload(raw)
     return res
   }
 
   async set(key: string, value: MessagePayload): Promise<any> {
```
Comment (MierenManz) [inline src/managers/messages.ts:37]: `set` can easily use `Promise<void>` or implicit return type because it does not return anything

**Review by MierenManz — CHANGES_REQUESTED** [review]
See my comments on suggested changes. Other than that. Looks good to me!
And really nice improvement. Goodjob DJ :D

File: src/managers/messages.ts (hunk 2 above)
Comment (DjDeveloperr) [reply src/managers/messages.ts:37]: linter goes brrr

File: src/managers/memberRoles.ts (hunk 1 above)
Comment (DjDeveloperr) [reply src/managers/memberRoles.ts:45]: _get can return undefined

File: src/cache/default.ts [hunk 3]
```diff
@@ -3,48 +3,57 @@ import type { ICacheAdapter } from './adapter.ts'
...
-      this.data[cacheName] = new Collection()
-      cache = this.data[cacheName]
+      cache = new Collection()
+      this.data.set(cacheName, cache)
     }
     cache.set(key, value)
     if (expire !== undefined)
       setTimeout(() => {
-        cache.delete(key)
+        cache?.delete(key)
```
[the remark this answers is not in the document]
Comment (DjDeveloperr) [reply src/cache/default.ts:22]: TS shows some error when i remove it lol

File: src/managers/messages.ts (hunk 2 above)
Comment (MierenManz) [reply src/managers/messages.ts:37]: then I'm quite curious why it is done somewhere else with another `set` method lol.
Eslint being stupid then I guess. Hopefully that can be changed soon

File: src/client/client.ts [hunk 4]
```diff
@@ -68,6 +68,10 @@ export interface ClientOptions {
   shard?: number
   /** ADVACNED: Shard count. */
   shardCount?: number | 'auto'
+  /** Whether to enable Zlib Compression or not (enabled by default) */
+  compress?: boolean
```
[the remark this answers is not in the document]
Comment (DjDeveloperr) [reply src/client/client.ts:72]: added

File: src/managers/messages.ts (hunk 2 above)
Comment (DjDeveloperr) [reply src/managers/messages.ts:37]: oh wait, I misread

Comment (DjDeveloperr) [reply src/managers/messages.ts]: fixed

File: src/managers/memberRoles.ts (hunk 1 above)
Comment (MierenManz) [reply src/managers/memberRoles.ts:45]: but doesn't
```ts
    if (payload === undefined) return 0
    else return payload.roles.length
```
Do the same as?
```ts
    if (payload === undefined) return 0
    return payload.roles.length
 ```

Comment (DjDeveloperr) [reply src/managers/memberRoles.ts:45]: well yes, i'll change that then

**Review by Helloyunho — APPROVED** [review]
lgtm!

**Pull request merged**